overlay/utils.sh

```diff
@@ -1,206 +1,283 @@
+      rerun-if-env-changed) ;&
+      warning)
+      ;;
+      rustc-link-search)
+        echo "-L $(printf '%q' "$val")" >> "$link_flags"
+        ;;
+      *)
+        if [ -e "$val" ]; then
+          local dep_file_target="$dep_files/DEP_$(upper "$cargo_links")_$(upper "$key")"
+          cp -r "$val" "$dep_file_target"
+          val=$dep_file_target
+        fi
+        printf 'DEP_%s_%s=%s\n' "$(upper "$cargo_links")" "$(upper "$key")" "$val" >> "$dep_keys"
+    esac
+  done < "$depinfo"
+}
+
+install_crate2() {
+  mkdir -p "$out/lib"
+  # shellcheck disable=SC2046
+  cp $(jq -rR 'fromjson?
+    | select(.reason == "compiler-artifact")
+    | .filenames
+    | map(select(test("\\.(a|rlib|so|dylib)$")))
+    | .[]' < .cargo-build-output) "$out/lib" 2> /dev/null || :
+
+  jq -rR 'fromjson?
+    | select(.reason == "build-script-executed")
+    | (.linked_libs | map("-l \(.)")) + (.linked_paths | map("-L \(.)"))
+    | .[]' < .cargo-build-output \
+    | tr '\n' ' ' >> "$out/lib/.link-flags"
+
+  mkdir -p "$out/bin"
+  # shellcheck disable=SC2046
+  cp $(jq -rR 'fromjson?
+    | select(.reason == "compiler-artifact")
+    | .executable
+    | select(.)' < .cargo-build-output) "$out/bin" 2> /dev/null || rmdir "$out/bin"
+
+  local out_dir
+  if out_dir="$(jq -rR 'fromjson? | select(.reason == "build-script-executed") | .out_dir' < .cargo-build-output)"; then
+    touch "$out/lib/.dep-keys"
+
+    # `links` envs aren't included in build output so we have to manually parse them.
+    if [[ -f "$out_dir/../output" ]]; then
+      grep -P '^cargo:(?!rerun-if-changed|rerun-if-env-changed|rustc-link-lib|rustc-link-search|rustc-flags|rustc-cfg|rustc-env|rustc-cdylib-link-arg|warning)' \
+        "$out_dir/../output" | while IFS= read -r line; do
+        [[ "$line" =~ cargo:([^=]+)=(.*) ]] || continue
         local key="${BASH_REMATCH[1]}"
         local val="${BASH_REMATCH[2]}"
+        printf 'DEP_%s_%s=%s\n' "$(upper "$cargo_links")" "$(upper "$key")" "$val" >> "$out/lib/.dep-keys"
+      done || :
+    fi
 
-        case $key in
-            rustc-link-lib) ;&
-            rustc-flags) ;&
-            rustc-cfg) ;&
-            rustc-env) ;&
-            rerun-if-changed) ;&
-            rerun-if-env-changed) ;&
-            warning)
-            ;;
-            rustc-link-search)
-                echo "-L" `printf '%q' $val` >>$link_flags
-                ;;
-            *)
-                if [ -e "$val" ]; then
-                    local dep_file_target=$dep_files/DEP_$(upper $cargo_links)_$(upper $key)
-                    cp -r "$val" $dep_file_target
-                    val=$dep_file_target
-                fi
-                printf 'DEP_%s_%s=%s\n' $(upper $cargo_links) $(upper $key) "$val" >>$dep_keys
-        esac
-    done
+    if [[ -d "$out_dir" ]] && { grep "$out_dir" "$out/lib/.dep-keys" || grep "$out_dir" "$out/lib/.link-flags"; }; then
+      local installed_out_dir="$out/build_output" # TODO: Change me
```

Comment:
I was not sure what this directory should be named. If this name looks good to you, we can remove the TODO.

Reply:
I'm honestly not sure what to make of this directory. What exactly is `$out/build_output` for? Why does it contain the contents of `bin` and `lib`?

Reply:
Would you mind elaborating what the purpose of this was, @trha?

Reply:
For future readers: it holds the contents of `$OUT_DIR` after a Cargo build script was run. These are often the result of build-time codegen or compiled non-Rust native libraries to be linked into the final crate.

Maybe we could rename it to `out_dir` in a subsequent PR to better clarify its purpose?

Reply:
@trha had an even better suggestion, in my mind, which is `build_script_output`. This should make it even more clear what this directory contains. Considering that it's an implementation detail of `cargo2nix`, it shouldn't matter too much what it's called, but that sounds clear enough to me.

Let's merge these changes in as-is, and I'll open a follow-up PR in the meantime.